### ISSUE TITLE
input filter

### ISSUE THREAD

**Issue by ianbale** [AddJam/ember-content-editable#9]
This is a great little control!

I have one request. Could you add a new function, much like stringInterpolator which filters what the user types in? eg. If I want a numeric value only, I can use stringInterpolator to strip out non-numeric characters, and only the numbers gets saved back to the model, but anything else I type in stays on the screen.  This new function should update the visible content rather than what is sent to the model. I guess you'd then still execute the stringInterpolator function afterwards to further manipulate the input before saving back to the model...

**Comment by chrissloey** [comment]
Hey, thanks!

I was able filter the input using the `key-press` event like this:

`{{content-editable value=name placeholder="Name" key-press=filter}}`

```
  filter(currentValue, event) {
    const key = event.which;
    if (key <= 48 || key >= 58) {
      event.preventDefault();
    }
  },
```

Works for both the input area and the value bound to the model :smile: if that solves your problem I'll close this.

I'll add an example of this to the README for now, because I think the stringInterpolator behaviour you mentioned is definitely not very intuitive - I'll have a look at improving that at some point this week, and maybe add support for things you would do on `<input>` like  `type="number"`.

**Comment by ianbale** [comment]
Thanks for that. I'll give it a whirl this evening (got a day job to do first!) and let you know later. 

I was also going to ask about limiting the length of the input. As long as we ignore someone pasting in content (which, thinking about it, might get past that filer too...), then I can check the length and if we've reached the max size then ignore the key. I guess if I filter on key-up so that if they are pasting, then the pasted content will be accessible to the filter, then I can see what might have been pasted and just change the value to whatever I want. Should work!

Might be nice to have a simple max-size param though.

**Comment by chrissloey** [comment]
Agreed - should definitely aim to make common use cases as easy as possible. I've been thinking the aim should be to support as much of what you'd get from `{{input}}` as possible, and `maxlength` is definitely [one of those things](http://guides.emberjs.com/v1.10.0/templates/input-helpers/).

Opened #10 to track the properties from input tags that should hopefully be possible to add. I should get a chance to implement some of this over the next week - happy to accept PRs too.

**Comment by ianbale** [comment]
That's not going to work for me quite as it is. I can, for example, add more than one "." in my number! But I guess I can fire off a function on key-press and if necessary overwrite the bound value. That should do the trick :-)

**Comment by chrissloey** [comment]
`maxlength` parameter now implemented (along with a few other things in #10) - there were a few edge cases with pasting and cursor positions etc. but I think everything is as it should be.

I think the current "number" type matches the behaviour of input. To support numbers like `1.23` I think we'd need to support the `step` attribute which might be outside of the scope of this component.

Is this ok to close now? :-)

**Comment by e-karma** [comment]
Well this could be exciting news!  Did not expect a solution to this problem after hearing this seemingly maddening experience.. https://vimeo.com/128728111 .  Did you manage to avoid the pitfalls encountered on that video?

Is it safe / ready to use in a real production app?  Going to spend some time on it.  Thank you!!!

**Comment by chrissloey** [comment]
Interesting video, thanks. I think it being ready for use in a real app depends on how you are using it. I started it while working on an app we're creating right now and I'd say it's production ready for what we need. This component is aiming to be as close to a drop in replacement for something like `{{input}}` or `{{textarea}}` as possible, with the value binding and events etc that those support.

It seems like what Intercom developed was a lot more complicated (e.g. the rich text editing stuff).  Without having used it for that I'm sure that some of the issues they faced will be there, but hopefully we can keep crossing them off the list as people come across them.

Let me know if it works out for you :smile:

**Comment by chrissloey** [comment]
I have come across some of the issues they mention.

The weird browser behaviour when newlines get inserted being one of the main ones - [there's a solution](https://github.com/addjam/ember-content-editable#extra-tags) if you don't care about preserving tags at all but currently nothing implemented beyond that.

I've also had to use the Selection APIs to manage cursor position when I implemented `maxlength`.

**Comment by e-karma** [comment]
Great, thanks for the hard-work and heads up!!  The more ambitious use-case I had envisioned was for an authenticated CMS user to have the ability to update site page / blog content on the actual site, not behind the CMS curtain.  That and a million other areas where its far superior to a separate "edit" screen if it can be avoided without exposing too much.   Does that use-case seem reasonably safe?

**Comment by chrissloey** [comment]
Our use case is pretty similar - editing a newsletter while keeping what you see as close to what will be sent as possible. I'd say it's worth trying out and hopefully we can fix any problems with the component if they arise.

**Comment by friksa** [comment]
We use it to edit a Rubrics (basically a large table) with each cell a content-editable component and everything works as advertised and very responsive!!

**Comment by e-karma** [comment]
Hey @chrissloey !  
Alright got almost everything working.  Thanks again!   1 issue however..

If using 'enter' event in component:
{{content-editable value=model.body tagName="h5" enter="saveEditable" type="text"}}
And when you click on text to edit it, change text then press enter, it ultimately skips to another line while enter event is fired to be handled by specified action. How can we make sure pressing enter will exit the 'input' state while keeping enter event active?

**Comment by chrissloey** [comment]
If you want to stop the enter event from creating a newline, I think the only way would be to make your action handler call `event.preventDefault()`? the action arguments should be `saveEditable(contentEditable, event)`.

**Comment by e-karma** [comment]
Thanks for trying, enter didn't work, but then I just realized clicking off works and looks great so i'm over it  until my todo list isn't so long.. Thanks again!

**Comment by chrissloey** [comment]
@e-karma 

Ran into this issue today and got it working, hopefully this will work for you too now. First problem was the enter event fired on `key-up` -  I think  `key-down` makes more sense for this since then we can actually prevent it, so I've updated that.

Other problem was calling blur() doesn't seem to work on most browsers for contenteditable elements.

Here's how I got it working:

`enter="endEditing"`

```
endEditing(contentEditable, event) {
  event.preventDefault();
  contentEditable.element.blur();
  window.getSelection().removeAllRanges();
}
```
